msal/src/Microsoft.Identity.Client/CacheV2/CacheManager.cs

```diff
@@ -0,0 +1,432 @@
+            {
+                // skipping writing to the cache, PK is empty
+                return null;
+            }
+
+            var credentialsToWrite = new List<Credential>();
+            long cachedAt = DateTime.UtcNow.Ticks; // todo: this is probably wrong
+
+            if (tokenResponse.HasRefreshToken)
+            {
+                credentialsToWrite.Add(
+                    Credential.CreateRefreshToken(
+                        homeAccountId,
+                        environment,
+                        clientId,
+                        cachedAt,
+                        tokenResponse.RefreshToken,
+                        string.Empty));
+            }
+
+            if (tokenResponse.HasAccessToken)
+            {
+                long expiresOn = tokenResponse.ExpiresOn.Ticks; // todo: this is probably wrong
+                long extendedExpiresOn = tokenResponse.ExtendedExpiresOn.Ticks; // todo: this is probably wrong
+
+                var accessToken = Credential.CreateAccessToken(
+                    homeAccountId,
+                    environment,
+                    realm,
+                    clientId,
+                    target,
+                    cachedAt,
+                    expiresOn,
+                    extendedExpiresOn,
+                    tokenResponse.AccessToken,
+                    string.Empty);
+                if (IsAccessTokenValid(accessToken))
+                {
+                    credentialsToWrite.Add(accessToken);
+                }
+            }
+
+            var idTokenJwt = tokenResponse.IdToken;
+            if (!idTokenJwt.IsEmpty)
+            {
+                credentialsToWrite.Add(
+                    Credential.CreateIdToken(
+                        homeAccountId,
+                        environment,
+                        realm,
+                        clientId,
+                        cachedAt,
+                        idTokenJwt.Raw,
+                        string.Empty));
+            }
+
+            var status = _storageManager.WriteCredentials(string.Empty, credentialsToWrite);
+            if (status.StatusType != OperationStatusType.Success)
+            {
+                // warning error writing to cache
+            }
+
+            // if id token jwt is empty, return null
+
+            string localAccountId = GetLocalAccountId(idTokenJwt);
+            var authorityType = GetAuthorityType();
+
+            var account = Microsoft.Identity.Client.CacheV2.Schema.Account.Create(
+                homeAccountId,
+                environment,
+                realm,
+                localAccountId,
+                authorityType,
+                idTokenJwt.PreferredUsername,
+                idTokenJwt.GivenName,
+                idTokenJwt.FamilyName,
+                idTokenJwt.MiddleName,
+                idTokenJwt.Name,
+                idTokenJwt.AlternativeId,
+                tokenResponse.RawClientInfo,
+                string.Empty);
+
```

Comment:
if some of these values are not returned, will that impact anything? for example, b2c idTokens might be different from aad idTokens as far as the claims that are returned. Also, why do we need all  these values for the account (ex. MiddleName)?

Reply:
If they come back empty, then they come back empty and we'll return empty fields for them.  This is all of the id token fields that were in C++ and we're going for functional consistency so I have them included here.